File: buy-sell-cooldown.js
// Time, Space - O(N), O(N)

/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    
    if(prices === null || prices.length < 2) {
        return 0;
    }
    let buy = new Array(prices.length);
    let sell = new Array(prices.length);    
    sell[0] = 0;
    buy[0] = -prices[0];
    
    for(let i=1;i<prices.length;i++) {
        
        if(i===1) {
            buy[i] = Math.max(buy[i-1], sell[i-1] - prices[i]);
            sell[i] = Math.max(sell[i-1], buy[i] + prices[i]);
        }
        else {
            buy[i] = Math.max(buy[i-1], sell[i-2] - prices[i]);
            sell[i] = Math.max(sell[i-1], buy[i] + prices[i]);  
        }
    }
    
    return sell[prices.length - 1];
};

File: buySellCooldown.java
// problem link : https://leetcode.com/problems/best-time-to-buy-and-sell-stock-with-cooldown/description/
// time complex : O(n)
// space Complex : O(n)

class Solution {
    int[] memo;
    int n ;
    int[] prices;
    public int maxProfit(int[] prices) {
        this.prices = prices;
        this.n = prices.length;
        this.memo = new int[n];
        Arrays.fill(memo, -1);
        return helper(0);
    }

    private int helper(int start){
        // base case
        if(start>=n)
            return 0;
        if(memo[start] != -1)
            return memo[start];

        // otherwise
        int currMax = 0 ; 
        int result = helper(start+1) ;
        for(int i = start; i<n ; i++){
            currMax = Math.max(currMax, prices[i]-prices[start]);
            result = Math.max(result, currMax + helper(i+2));
        }
        memo[start] = result;
        return result;
    }
}
